File: tool/worker.mjs
// re/tool/worker.mjs — chạy login chain cho 1 account (1 cửa sổ CMD), in ✓/✗ từng bước,
//   thành công thì FETCH + in INFO chi tiết (follower/following/video/likes...). Bám note 26.
//   Account: env ACCOUNT hoặc argv[2] = "user|pass|email|mailpass[|did|iid|openudid|cdid|gaid]".
//   Proxy:   env PROXY_URL (mỗi cửa sổ 1 IP). Signer: env METASEC_ORACLE (genuine) | mặc định unidbg offline.
import '../src/net.mjs';            // side-effect: proxy egress từ PROXY_URL
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import readline from 'node:readline';
import { fileURLToPath } from 'node:url';
import { registerDevice, dsign } from '../src/device.mjs';
import { warmup, preCheck, userLogin, passportCall, seedCookies, getJar, cookieHdr, enc, JAR } from '../src/login.mjs';
import { challenges, authSend, authVerify, newPseudoId } from '../src/aaas.mjs';
import { callAuthed } from '../src/session.mjs';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const OUT = path.join(HERE, 'out');
const DEV = path.join(HERE, 'devices');                       // mỗi account 1 device_id BỀN
const safeName = (s) => String(s).replace(/[^\w.-]/g, '_').slice(0, 60);

// ── error model: StepError chỉ đúng HÀM hỏng khi TikTok update ──
const HINTS = {
  '*7': 'ec7 = velocity/rate-limit theo device_id + IP-register (note 26). KHÔNG phải X-Argus/recipe. '
      + 'Mint device trên IP residential SẠCH + login CÙNG IP; throttle đã trip = chờ giờ.',
  '*1105': 'ec1105 = captcha — device forge/untrusted hoặc IP bẩn. Cần trusted device (dsign s=1) + IP sạch.',
  '*2135': 'ec2135 = account bị-cờ (ở user_login = đúng kỳ vọng → đi tiếp aaas).',
  'sign': 'signer hỏng — METASEC_ORACLE (genuine) chết, hoặc unidbg offline chưa compile/JDK21. '
        + 'Nhớ: login account CŨ thường cần X-Argus GENUINE (phone-oracle), offline chỉ đủ format.',
  'register_device': 'device_register không trả device_id — schema đổi hoặc bị chặn (IP/fingerprint).',
  'dsign': 'dsign http≠200 — device bị ban hoặc device-guard đổi.',
  'challenges': 'không có factor type=2 (email) — luồng verify đổi, hoặc ticket/pseudo_id sai.',
  'relogin': 'relogin #7 fail — header x-tt-passport-ticket + d_ticket + cookie strip 5-key (note 26).',
};
const hintFor = (step, ec) => HINTS['*' + ec] || HINTS[step] || null;

class StepError extends Error {
  constructor(step, layer, o = {}) {
    super(`[${layer}] ${step}: ec=${o.ec} http=${o.http}`);
    Object.assign(this, { step, layer, ...o });
    this.hint = o.hint || hintFor(step, o.ec);
  }
  report() {
    const L = [`\n✗ ${this.step}  [${this.layer}]`];
    if (this.endpoint) L.push(`    endpoint ${this.endpoint}`);
    const m = [];
    if (this.http != null) m.push(`http=${this.http}`);
    if (this.ec != null) m.push(`ec=${this.ec}`);
    if (m.length) L.push('    ' + m.join('  '));
    if (this.msg) L.push(`    msg: ${this.msg}`);
    if (this.hint) L.push(`    hint: ${this.hint}`);
    if (this.raw != null) L.push('    raw: ' + (typeof this.raw === 'string' ? this.raw : JSON.stringify(this.raw)).slice(0, 600));
    return L.join('\n');
  }
}

const ok = (s, d = '') => console.log(`✓ ${s}` + (d ? `   ${d}` : ''));

// gọi 1 step, throw StepError (attribute đúng hàm) nếu lỗi hạ tầng (network/sign)
async function run(step, fn) {
  try { return await fn(); }
  catch (e) {
    if (e instanceof StepError) throw e;
    const msg = e.message || String(e);
    let ohost = ''; try { ohost = process.env.METASEC_ORACLE ? new URL(process.env.METASEC_ORACLE).host : ''; } catch {}
    // signer: lỗi nhắc argus/gorgon/oracle-sign, unidbg/java/mvn (offline), hoặc trúng host oracle
    const isSigner = /argus|gorgon|khronos|ladon|oracle sign|unidbg|\bjava\b|\bmvn\b|jdk|Harness/i.test(msg) || (ohost && msg.includes(ohost));
    if (isSigner) throw new StepError(step, 'SIGN', { msg, hint: HINTS.sign });
    throw new StepError(step, 'NET', { msg, hint: 'Lỗi mạng — check PROXY_URL còn sống + signer (METASEC_ORACLE/unidbg) còn chạy + host reachable.' });
  }
}

// business ec: success → 'success'; ec ∈ allow → ec; còn lại throw StepError
function checkEc(r, step, layer, endpoint, allow = []) {
  if (r.j?.message === 'success') return 'success';
  const ec = r.j?.data?.error_code ?? r.j?.message;
  if (allow.includes(ec)) return ec;
  throw new StepError(step, layer, { endpoint, http: r.status, ec, msg: r.j?.data?.description || r.j?.message, raw: r.j });
}

// strip cookie 5-key cho relogin (note 26)
const strip5 = () => { const J = getJar(); return ['store-idc', 'tt-target-idc', 'odin_tt', 'd_ticket', 'msToken'].filter(k => J[k] != null).map(k => `${k}=${J[k]}`).join('; '); };

// ── đọc mã email: RE_CODE > mail.tm > stdin ──
async function mailtmRead(address, password, timeoutMs = 120000) {
  const tok = await (await fetch('https://api.mail.tm/token', { method: 'POST', headers: { 'content-type': 'application/json' }, body: JSON.stringify({ address, password }) })).json();
  if (!tok.token) throw new Error('mail.tm auth fail');
  const H = { authorization: 'Bearer ' + tok.token };
  const seen = new Set(); const deadline = Date.now() + timeoutMs;
  const grab = (t) => (t && (t.match(/\b(\d{6})\b/) || t.match(/\b(\d{4,5})\b/)) || [])[1] || null;
  while (Date.now() < deadline) {
    const lst = (await (await fetch('https://api.mail.tm/messages', { headers: H })).json())['hydra:member'] || [];
    for (const m of lst) {
      if (seen.has(m.id)) continue; seen.add(m.id);
      let c = grab(m.subject);
      if (!c) { try { const full = await (await fetch('https://api.mail.tm/messages/' + m.id, { headers: H })).json(); c = grab(full.text) || grab(full.subject); } catch {} }
      if (c) return c;
    }
    await new Promise(r => setTimeout(r, 4000));
  }
  return null;
}
function promptStdin(q) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise(res => rl.question(q, a => { rl.close(); res((a || '').trim()); }));
}
async function readCode(acc) {
  if (process.env.RE_CODE) return process.env.RE_CODE;
  if (acc.email && acc.mailpass) { try { const c = await mailtmRead(acc.email, acc.mailpass); if (c) return c; } catch {} }
  return promptStdin(`Nhập mã verify gửi tới ${acc.email || '(email)'}: `);
}

function parseAccount(line) {
  const f = String(line).split('|').map(s => s.trim());
  return { username: f[0], password: f[1], email: f[2] || '', mailpass: f[3] || '', did: f[4] || '', iid: f[5] || '', openudid: f[6] || '', cdid: f[7] || '', gaid: f[8] || '' };
}

// ── in INFO chi tiết + lưu ──
async function showInfo(uid, session) {
  const prof = await callAuthed(session, '/aweme/v1/user/profile/self/', { extraQuery: { user_id: uid } }).catch(e => ({ status: 'ERR', j: null, err: e.message }));
  const u = prof.j?.data?.user || prof.j?.user || {};
  let acc = {};
  try { const ai = await callAuthed(session, '/passport/account/info/v2/', { extraQuery: { scene: 'normal' } }); acc = ai.j?.data || {}; } catch {}
  const info = {
    ok: !!(u.uid || u.nickname), uid, nickname: u.nickname ?? null, unique_id: u.unique_id ?? null,
    signature: u.signature ?? null, region: u.region ?? null,
    follower_count: u.follower_count ?? null, following_count: u.following_count ?? null,
    aweme_count: u.aweme_count ?? null, total_favorited: u.total_favorited ?? null,
    favoriting_count: u.favoriting_count ?? null, create_time: u.create_time ?? null,
    email: acc.email ?? null, mobile: acc.mobile ?? null, has_password: acc.has_password ?? null,
  };
  console.log('\n════════════ THÔNG TIN TÀI KHOẢN ════════════');
  console.log(`  uid          : ${uid}`);
  console.log(`  nickname     : ${info.nickname}`);
  console.log(`  unique_id (@): ${info.unique_id}`);
  console.log(`  follower     : ${info.follower_count}   (người theo dõi)`);
  console.log(`  following    : ${info.following_count}   (đang theo dõi)`);
  console.log(`  video        : ${info.aweme_count}`);
  console.log(`  ❤ nhận (like): ${info.total_favorited}   (total_favorited)`);
  console.log(`  ❤ đã thả     : ${info.favoriting_count}   (favoriting_count)`);
  console.log(`  region       : ${info.region}`);
  console.log(`  email        : ${info.email}`);
  console.log(`  has_password : ${info.has_password}`);
  console.log('══════════════════════════════════════════════');
  try { fs.mkdirSync(OUT, { recursive: true }); fs.writeFileSync(path.join(OUT, `${uid}.json`), JSON.stringify({ ...info, cookie: session.cookie, ts: Date.now() }, null, 2)); console.log(`  (đã lưu ${path.join('out', uid + '.json')})`); } catch {}
  return info;
}

async function main() {
  const line = process.env.ACCOUNT || process.argv[2];
  if (!line || line.split('|').length < 2) { console.error('cần ACCOUNT="user|pass|email|mailpass"'); process.exit(2); }
  const acc = parseAccount(line);
  const u = acc.username, pw = acc.password;
  console.log('── re/tool login chain ──');
  console.log(`   account=${u}  signer=${process.env.METASEC_ORACLE || 'unidbg-offline'}  proxy=${process.env.PROXY_URL ? 'ON' : 'off'}`);

  try {
    // 02 device: MỖI ACCOUNT 1 device_id RIÊNG + BỀN (lưu re/tool/devices/<user>.json).
    //    ưu tiên: device provided ở account.txt > device đã lưu (tái dùng) > register mới rồi LƯU.
    let dev;
    const devFile = path.join(DEV, safeName(u) + '.json');
    if (acc.did && acc.iid) {
      dev = { device_id: acc.did, install_id: acc.iid, id: { openudid: acc.openudid, cdid: acc.cdid, google_aid: acc.gaid || crypto.randomUUID() }, cookies: {} };
      ok('02 device (provided)', `did=${dev.device_id} iid=${dev.install_id}`);
    } else if (fs.existsSync(devFile)) {
      dev = JSON.parse(fs.readFileSync(devFile, 'utf8')); dev.cookies = {};
      ok('02 device (reuse)', `did=${dev.device_id} iid=${dev.install_id} (đã lưu → giữ device riêng)`);
    } else {
      dev = await run('register_device', () => registerDevice());
      if (!dev.device_id) throw new StepError('register_device', 'DEVICE', { endpoint: '/service/2/device_register/', raw: dev.raw });
      try { fs.mkdirSync(DEV, { recursive: true }); fs.writeFileSync(devFile, JSON.stringify({ device_id: dev.device_id, install_id: dev.install_id, new_user: dev.new_user, id: dev.id, profile: process.env.RE_PROFILE ?? null, ts: Date.now() }, null, 2)); } catch {}
      ok('02 register_device (mới)', `did=${dev.device_id} iid=${dev.install_id} new_user=${dev.new_user} → đã lưu devices/${safeName(u)}.json`);
    }

    // 03 dsign + guards
    const d = await run('dsign', () => dsign(dev));
    ok('03 dsign+guards', `s=${d.s} ts_sign=${d.ts_sign ? 'yes' : 'no'}`);

    // 04 seed cookie odin_tt
    seedCookies(dev.cookies); seedCookies(d.cookies);
    ok('04 seed_cookies', `jar=${Object.keys(getJar())}`);

    // 05 warmup (best-effort)
    await run('warmup', () => warmup(dev, d)).catch(() => {});
    ok('05 warmup', `jar=${Object.keys(getJar())}`);

    // 06 pre_check (best-effort)
    try { const pc = await preCheck(u, dev, d); ok('06 pre_check', `status=${pc.status} ec=${pc.j?.data?.error_code} msg=${pc.j?.message}`); }
    catch (e) { console.log('  (06 pre_check bỏ qua) ' + e.message); }

    // 07 user_login → success | 2135 | ec7
    const lg = await run('user_login', () => userLogin(u, pw, dev, d));
    const res = checkEc(lg, 'user_login', 'LOGIN', '/passport/user/login/', [2135]);
    if (res === 'success') {
      ok('07 user_login', 'ĐĂNG NHẬP THẲNG (account chưa bị cờ)');
      const uid = lg.j?.data?.user_id_str || String(lg.j?.data?.user_id || '');
      ok('DONE', `uid=${uid} session_key=${(lg.j?.data?.session_key || '').slice(0, 12)}…`);
      await showInfo(uid, { cookie: cookieHdr(), deviceId: dev.device_id, iid: dev.install_id, xtt: lg.xtt || '' });
      return finish();
    }
    // 2135 branch
    const dc = lg.dc || {};
    const ticket = dc.passport_ticket;
    const pid = (dc.extra?.[0]?.pseudo_id) || newPseudoId();
    if (!ticket) throw new StepError('user_login', 'LOGIN', { endpoint: '/passport/user/login/', http: lg.status, ec: 2135, raw: dc, hint: '2135 nhưng thiếu passport_ticket ở header x-tt-verify-idv-decision-conf.' });
    ok('07 user_login', `2135 (bị cờ, đúng kỳ vọng) ticket=${ticket.slice(0, 10)}… pid=${pid.slice(0, 10)}…`);

    // 08 challenges → factor type=2
    const ch = await run('challenges', () => challenges(dev, d, ticket));
    checkEc(ch, 'challenges', 'AAAS', '/passport/aaas/challenges/', [0, undefined]);
    const factors = ch.j?.data?.challenges || ch.j?.challenges || [];
    const types = factors.map(c => c.type);
    if (!types.includes(2)) throw new StepError('challenges', 'AAAS', { endpoint: '/passport/aaas/challenges/', http: ch.status, raw: ch.j, hint: `không có factor type=2 (email). types=${JSON.stringify(types)}.` });
    ok('08 challenges', `factors=${JSON.stringify(types)} (2=email)`);

    // 09 auth_send action=3
    const se = await run('auth_send', () => authSend(dev, d, ticket, pid));
    checkEc(se, 'auth_send', 'AAAS', '/passport/aaas/authenticate/');
    ok('09 auth_send', 'server đã gửi mã tới email');

    // 10 đọc mã email
    const code = await readCode(acc);
    if (!code) throw new StepError('read_code', 'EMAIL', { hint: 'Không lấy được mã (timeout/không tới). Set RE_CODE hoặc nhập tay.' });
    ok('10 read_code', `code=${code}`);

    // 11 auth_verify action=4 → d_ticket
    const vf = await run('auth_verify', () => authVerify(dev, d, ticket, pid, code));
    checkEc(vf, 'auth_verify', 'AAAS', '/passport/aaas/authenticate/');
    ok('11 auth_verify', `verified d_ticket=${vf.d_ticket ? 'yes' : 'NO(!)'}`);

    // 12 relogin #7 → session
    if (vf.d_ticket) JAR['d_ticket'] = vf.d_ticket;
    const rl = await run('relogin', () => passportCall(dev, d, '/passport/user/login/', {
      params: { password: enc(pw), account_sdk_source: 'app', multi_login: '1', mix_mode: '1', username: enc(u) },
      extra: { 'x-tt-retry-by-x-tt-verify-idv-decision-conf': '1', 'x-tt-passport-ticket': ticket },
      cookieOverride: strip5(),
    }));
    checkEc(rl, 'relogin', 'LOGIN', '/passport/user/login/');
    const uid = rl.j?.data?.user_id_str || String(rl.j?.data?.user_id || '');
    ok('12 relogin', `uid=${uid} session_key=${(rl.j?.data?.session_key || '').slice(0, 12)}…`);
    console.log(`✓ DONE  session_key=${rl.j?.data?.session_key}  user_id=${uid}`);
    await showInfo(uid, { cookie: cookieHdr(), deviceId: dev.device_id, iid: dev.install_id, xtt: rl.xtt || lg.xtt || '' });
    return finish();

  } catch (e) {
    if (e instanceof StepError) { console.log(e.report()); appendFail(u, e); }
    else { console.log(`\n✗ LỖI KHÔNG XÁC ĐỊNH: ${e.stack || e.message}`); appendFail(u, { step: 'unknown', ec: '', report: () => e.message }); }
    return finish(1);
  }
}

function appendFail(user, e) {
  try { fs.mkdirSync(OUT, { recursive: true }); fs.appendFileSync(path.join(OUT, 'fail.txt'), `${user}\t${e.step}\tec=${e.ec ?? ''}\n`); } catch {}
}
function finish(code = 0) {
  // giữ cửa sổ mở để đọc kết quả (trừ khi NO_PAUSE)
  if (process.env.NO_PAUSE) return process.exit(code);
  return promptStdin('\n[Enter để đóng cửa sổ] ').then(() => process.exit(code));
}

main();

File: py/tool/worker.py
"""re/py/tool/worker.py — chạy login-2135 chain cho 1 account (mỗi account 1 device BỀN), in ✓/✗ từng bước,
   thành công thì FETCH + in INFO (follower/following/video/likes...). Port re/tool/worker.mjs.
  ACCOUNT env | argv[1] = "user|pass|email|mailpass[|did|iid|openudid|cdid|gaid]"
  PROXY_URL  : mỗi tiến trình 1 IP (net.py egress).  SIGNER_URL/METASEC_ORACLE: signer.
  FOLLOW=<uniqueId> [FOLLOW_TYPE=1|0] : sau login → follow + verify STUCK/shadow.
  SAVE_SESSION=<path> : lưu session ra file (tái dùng bằng check_follow, KHÔNG re-login).
"""
import os
import sys
import json
import uuid

HERE = os.path.dirname(os.path.abspath(__file__))
REPY = os.path.dirname(HERE)
sys.path.insert(0, REPY)
try:
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')
except Exception:
    pass

import net  # noqa: E402  (side-effect: proxy egress từ PROXY_URL)
import ui  # noqa: E402  (màu + box console)
import session as sess_mod  # noqa: E402
import follow as follow_mod  # noqa: E402
from net import now_ms  # noqa: E402
from chain import run_login_chain  # noqa: E402
from run import parse_account, make_reader  # noqa: E402

OUT = os.path.join(HERE, 'out')
DEV = os.path.join(HERE, 'devices')   # mỗi account 1 device_id BỀN


def safe_name(s):
    return ''.join(c if (c.isalnum() or c in '.-_') else '_' for c in str(s))[:60]


def append_fail(user, step, ec):
    try:
        os.makedirs(OUT, exist_ok=True)
        with open(os.path.join(OUT, 'fail.txt'), 'a', encoding='utf-8') as f:
            f.write(f'{user}\t{step}\tec={ec or ""}\n')
    except Exception:
        pass


def finish(code=0):
    if os.environ.get('NO_PAUSE'):
        return code
    try:
        input('\n[Enter để đóng cửa sổ] ')
    except EOFError:
        pass
    return code


def show_info(session):
    """FETCH + in INFO chi tiết + lưu out/<uid>.json."""
    uid = session['uid']
    u = {}
    try:
        prof = sess_mod.call_authed(session, '/aweme/v1/user/profile/self/', extra_query={'user_id': uid})
        pj = prof.get('j') or {}
        u = (pj.get('data') or {}).get('user') or pj.get('user') or {}
    except Exception:
        pass
    acc = {}
    try:
        ai = sess_mod.call_authed(session, '/passport/account/info/v2/', extra_query={'scene': 'normal'})
        acc = (ai.get('j') or {}).get('data') or {}
    except Exception:
        pass
    info = {
        'ok': bool(u.get('uid') or u.get('nickname')), 'uid': uid,
        'nickname': u.get('nickname'), 'unique_id': u.get('unique_id'), 'signature': u.get('signature'),
        'region': u.get('region'), 'follower_count': u.get('follower_count'),
        'following_count': u.get('following_count'), 'aweme_count': u.get('aweme_count'),
        'total_favorited': u.get('total_favorited'), 'favoriting_count': u.get('favoriting_count'),
        'create_time': u.get('create_time'), 'email': acc.get('email'), 'mobile': acc.get('mobile'),
        'has_password': acc.get('has_password'),
    }
    print()
    ui.box_top(info['unique_id'] or str(uid), tag='SESSION ✓')
    ui.row('nickname', info['nickname'])
    ui.row('follower', info['follower_count'], 'following', info['following_count'])
    ui.row('video', info['aweme_count'], '❤ nhận', info['total_favorited'])
    ui.row('❤ đã thả', info['favoriting_count'], 'region', info['region'])
    ui.row('email', info['email'], 'password', 'có ✓' if info['has_password'] else '—')
    ui.row('uid', uid)
    ui.box_bot()
    try:
        os.makedirs(OUT, exist_ok=True)
        with open(os.path.join(OUT, f'{uid}.json'), 'w', encoding='utf-8') as f:
            json.dump({**info, 'cookie': session.get('cookie'), 'ts': now_ms()}, f, ensure_ascii=False, indent=2)
        print(f"  (đã lưu out/{uid}.json)")
    except Exception:
        pass
    return info


def load_persist_device(user, fields):
    """Ưu tiên: device provided ở account.txt > device đã lưu (tái dùng) > None (chain sẽ register mới)."""
    did = fields[4] if len(fields) > 4 else ''
    iid = fields[5] if len(fields) > 5 else ''
    if did and iid:
        idn = {'openudid': fields[6] if len(fields) > 6 else '',
               'cdid': fields[7] if len(fields) > 7 else '',
               'google_aid': (fields[8] if len(fields) > 8 else '') or str(uuid.uuid4())}
        print(f'✓ 02 device (provided)   did={did} iid={iid}')
        return {'device_id': did, 'install_id': iid, 'id': idn}
    dev_file = os.path.join(DEV, safe_name(user) + '.json')
    if os.path.exists(dev_file):
        with open(dev_file, encoding='utf-8') as f:
            s = json.load(f)
        print(f"✓ 02 device (reuse)      did={s['device_id']} iid={s['install_id']} (giữ device riêng)")
        return {'device_id': s['device_id'], 'install_id': s['install_id'], 'id': s.get('id') or {}}
    return None


def save_persist_device(user, dev):
    try:
        os.makedirs(DEV, exist_ok=True)
        with open(os.path.join(DEV, safe_name(user) + '.json'), 'w', encoding='utf-8') as f:
            json.dump({'device_id': dev['device_id'], 'install_id': dev['install_id'],
                       'new_user': dev.get('new_user'), 'id': dev.get('id'), 'ts': now_ms()}, f)
    except Exception:
        pass


def main():
    line = os.environ.get('ACCOUNT') or (sys.argv[1] if len(sys.argv) > 1 else None)
    if not line or len(line.split('|')) < 2:
        print('cần ACCOUNT="user|pass|email|mailpass"')
        return 2
    fields = [x.strip() for x in line.split('|')]
    acc = parse_account(line)
    u = acc['username']
    ui.head(f'LOGIN  {u}')
    print(f"   {ui.DIM}signer={os.environ.get('SIGNER_URL') or os.environ.get('METASEC_ORACLE') or 'signer-bridge'}"
          f"  proxy={'ON' if os.environ.get('PROXY_URL') else 'off'}{ui.RESET}")

    dev = load_persist_device(u, fields)
    r = run_login_chain(acc, read_code=make_reader(acc), dev=dev)
    if dev is None and r.get('dev'):      # freshly registered → LƯU device để tái dùng
        save_persist_device(u, r['dev'])

    if not r['ok']:
        append_fail(u, r.get('step'), getattr(r.get('error', None), 'ec', ''))
        return finish(1)

    s = r['session']
    s['user'] = u
    show_info(s)

    # follow tùy chọn (write-op)
    target = os.environ.get('FOLLOW')
    if target and r.get('dev') and r.get('d'):
        fr = follow_mod.follow_and_verify(r['dev'], r['d'], target, s.get('xtt', ''),
                                          cookie=s.get('cookie'), type=os.environ.get('FOLLOW_TYPE', '1'))
        if fr.get('ok'):
            v = fr['verdict']
            tag = (f"{ui.GREEN}✅ STUCK (follow thật){ui.RESET}" if v == 'stuck'
                   else (f"{ui.YELLOW}⚠️ shadow-drop{ui.RESET}" if v == 'shadow' else v))
            print(f"\n{ui.CYAN}➤ FOLLOW{ui.RESET} @{fr['target']}: sc={fr['follow_sc']} "
                  f"follow_status {fr['follow_status_before']}→{fr['follow_status_after']} → {tag}")
        else:
            print(f"\n{ui.CYAN}➤ FOLLOW{ui.RESET} @{target}: {ui.YELLOW}{fr.get('reason', 'fail')}{ui.RESET}")

    # lưu session tùy chọn
    sp = os.environ.get('SAVE_SESSION')
    if sp and r.get('dev') and r.get('d'):
        try:
            sess_mod.save_session(s, r['dev'], r['d'], sp)
            print(f'  (đã lưu session → {sp})')
        except Exception as e:
            print(f'  (lưu session lỗi: {e})')
    return finish(0)


if __name__ == '__main__':
    sys.exit(main())
